### DISCUSSION TITLE
Variable replacement with empty value

### DISCUSSION THREAD

**Discussion by sylvainmouquet** [apple/pkl#864] [General]
Hello,

I use this code : `read("env:FOO")`
If the env variable contains `$bar-bar` the `$bar` is replaced by empty value. 

To help the developer to track bad defintioon, we could add a warning when a variable is defined without a value.

**Comment by HT154** [comment]
Unfortunately, this is not how environment variables work in general. By the time the environment variable data makes it into Pkl's process it has already been "expanded" and Pkl has no way of detecting that `-bar` is missing something because `$bar` was undefined. You can see this with any process, eg.
```console
❯ export foo="$bar-bar" # $bar is undefined
❯ bash # start a new shell, inheriting the environment
$ env | grep foo
foo=-bar
```

Some shells like `bash` and `zsh` support an alternate mode that makes usage of an undefined variable an error. In `bash`, you can enable this with `set -u`. That said, this mode of operation can break many scripts that rely on the default behavior of undefined variables.

If you're expecting the value to match a specific pattern, you may be able to at least do _some_ validation in Pkl using type constraints, eg.
```pkl
local foo: String(matches(Regex("<some regex>"))) = read("env:FOO")
```
